coffea/nanoevents/schemas/edm4hep.py

```diff
@@ -0,0 +1,199 @@
+import re
+
+from coffea.nanoevents.schemas.base import BaseSchema, nest_jagged_forms, zip_forms
+
+_base_collection = re.compile(r".*[\#\/]+.*")
+_trailing_under = re.compile(r".*_[0-9]")
+
+
+class EDM4HEPSchema(BaseSchema):
+    """EDM4HEP schema builder
+
+    The EDM4HEP schema is built from all branches found in the supplied file,
+    based on the naming pattern of the branches. There are two steps of to the
+    generation of array collections:
+
+    - Objects with vector-like quantities (momentum, coordinate points) in the
+      TreeMaker ntuples are stored using ROOT PtEtaPhiEVectors and XYZPoint
+      classes with maximum TTree splitting. These variable branches are grouped
+      into a single collection with the original object name, with the
+      corresponding coordinate variables names mapped to the standard variable
+      names for coffea.nanoevents.methods.vector behaviors. For example:
+      - The "Jets" branch in a TreeMaker Ntuple branch stores 'PtEtaPhiEVector's
+        corresponding to the momentum of AK4 jets. The resulting collection after
+        this first step would contain the vector variables in the form of
+        Jets.pt, Jets.eta, Jets.phi, Jets.energy, and addition vector quantities
+        (px) can be accessed via the usual vector behavior methods.
+      - The "PrimaryVertices" branch in a TreeMaker Ntuple branch stores
+        'XYZPoint's corresponding to the coordinates of the primary vertices, The
+        resulting collection after this first step wold contain the coordinate
+        variables in the form of PrimaryVertices.x, PrimaryVertices.y,
+        PrimaryVertices.z.
+
+    - Extended quantities of physic objects are stored in the format
+      <Object>_<variable>, such as "Jets_jecFactor". Such variables will be
+      merged into the collection <Object>, so the branch "Jets_jetFactor" will be
+      access to in the array format as "Jets.jecFactor". An exception to the
```

Comment:
In "podio speak" these are *`VectorMembers`* and they are used in several of the datatypes of [edm4hep](https://github.com/key4hep/EDM4hep/blob/1145e20b9ee238e1110f643232cabd061646582e/edm4hep.yaml).

Note that there are also *`OneToOneRelations`* and *`OneToManyRelations`* which are currently still stored using the convenction `<collection-name>#<index>`. These are the branches you would need to use to resolve e.g. mother/daughter relations for MCParticles. The `<index>` is currently somewhat "magic" but **this is going to change** with AIDASoft/podio#405, where the new schema of the branch names will be `_<collection-name>_<name>` for both `VectorMembers` and the relations. Let me know if you want to wait for that or whether you want some information about the magic.

I am not sure how general you want to make this in the end, but we could potentially help with some code generation on the podio side if you think that would be in any way useful.

Reply:
@tmadlener Thanks for the info. Sorry, the documentation is for another dataformat type - I'll clean it up next week. In any case I can always find the constituents of a given physics object and wrap them together.

So far as code generation - once we work it out here it could be nice to move it properly to `podio` (though that's not a python package but rather a generator for one)... Could envision some CI to put something on PyPI. Though that may not be well advised since what's being generated here is the schema as opposed to the implementation (which in our case is handled by more general functions + an embedded domain specific language that does know how to handle cross references if you give it the source/mapping/target). It may be best that it stays here, it only needs to be done once anyway since it gets distributed with coffea? We can work it out in time.

I think in the mean time if you could help us with the "magic" cross references so we know what we're supposed to be referencing that would be really useful! We can add in the better-named xref branches when they're there and files with it are available.

Reply:
I agree on the code generation. Let's see how this works out and how frequent changes would need to be.

So the "magic" is effectively that the indices get generated sequentially in the order the relations appear in the yaml definition file, **starting with `OneToManyRelations` and then continuing with `OneToOneRelations`**. So as an example the [`edm4hep::ReconstructedParticle`](https://github.com/key4hep/EDM4hep/blob/1145e20b9ee238e1110f643232cabd061646582e/edm4hep.yaml#L409-L437) has
```yaml
    OneToOneRelations:
      - edm4hep::Vertex          startVertex    //start vertex associated to this particle
      - edm4hep::ParticleID      particleIDUsed //particle Id used for the kinematics of this particle
    OneToManyRelations:
      - edm4hep::Cluster               clusters     //clusters that have been used for this particle.
      - edm4hep::Track                 tracks       //tracks that have been used for this particle.
      - edm4hep::ReconstructedParticle particles    //reconstructed particles that have been combined to this particle.
      - edm4hep::ParticleID            particleIDs  //particle Ids (not sorted by their likelihood)
```

and on file the branches will be (assuming the collection name is `"reco"`)
- `reco#0` for the related Clusters
- `reco#1` for the related Tracks
- ...
- `reco#4` for the (single) startVertex
- `reco#5` for the (single) particleIDUsed

Reply:
Indeed that's a little bit hacky to keep working since there's no clear way to keep track of the schema over time.

The next version sounds much more tractable in the long run!